package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,590 @@
+        dirs
+    );
+    ssh_exec(server, &cmd, false)
+}
+
+fn single_server_install(
+    config: &Config,
+    artifact_dir: &Path,
+    install_dir: &Path,
+    pkg_dir: &str,
+    builder: &Server,
+    server_name: &str,
+) -> Result<()> {
+    let server = &config.servers[server_name];
+
+    copy_package_artifacts_to_staging(config, pkg_dir, builder, server)?;
+    copy_omicron_package_binary_to_staging(config, builder, server)?;
+    copy_package_manifest_to_staging(config, builder, server)?;
+    run_omicron_package_from_staging(
+        config,
+        server,
+        &artifact_dir,
+        &install_dir,
+    )?;
+    copy_overlay_files_to_staging(
+        config,
+        pkg_dir,
+        builder,
+        server,
+        server_name,
+    )?;
+    install_overlay_files_from_staging(config, server, &install_dir)?;
+    restart_services(server)
+}
+
+// Copy package artifacts as a result of `omicron-package package` from the
+// builder to the deployment server staging directory.
+fn copy_package_artifacts_to_staging(
+    config: &Config,
+    pkg_dir: &str,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let cmd = format!(
+        "rsync -avz -e 'ssh -o StrictHostKeyChecking=no' \
+                    --exclude overlay/ {} {}@{}:~/{}",
+        pkg_dir,
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn copy_omicron_package_binary_to_staging(
+    config: &Config,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let mut bin_path = PathBuf::from(&config.builder.omicron_path);
+    bin_path.push("target/debug/omicron-package");
+    let cmd = format!(
+        "rsync -avz {} {}@{}:~/{}",
+        bin_path.to_string_lossy(),
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn copy_package_manifest_to_staging(
+    config: &Config,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let mut path = PathBuf::from(&config.builder.omicron_path);
+    path.push("package-manifest.toml");
+    let cmd = format!(
+        "rsync {} {}@{}:~/{}",
+        path.to_string_lossy(),
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn run_omicron_package_from_staging(
+    config: &Config,
+    destination: &Server,
+    artifact_dir: &Path,
+    install_dir: &Path,
+) -> Result<()> {
+    let mut deployment_src = PathBuf::from(&config.deployment.staging_dir);
+    deployment_src.push(&artifact_dir);
+
+    // Run `omicron-package install` on the deployment server
+    let cmd = format!(
+        "cd ~/{} && pfexec ./omicron-package install --in ~/{} --out {}",
```

Comment:
Same thing here. Is this `$HOME` ensured to be expanded correctly?

Reply:
As far as I can tell it always is. If you feel strongly I can change it.

Reply:
Changed to use $HOME.